// @aws-cdk/region-info/lib/fact.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FactName = exports.Fact = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_entities_1 = require("./aws-entities");
const CONSTRUCT_ERROR_SYMBOL = Symbol.for('@aws-cdk/core.SynthesisError');
/**
 * An error thrown when a region fact operation fails.
 */
class RegionFactError extends Error {
    constructor(msg) {
        super(msg);
        Object.setPrototypeOf(this, RegionFactError.prototype);
        Object.defineProperty(this, CONSTRUCT_ERROR_SYMBOL, { value: true });
        this.name = 'RegionFactError';
    }
}
/**
 * A database of regional information.
 */
class Fact {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/region-info.Fact", version: "2.235.0" };
    /**
     * @returns the list of names of AWS Regions for which there is at least one registered fact. This
     *          includes Regions defined in AWS_REGIONS plus custom defined regions.
     */
    static get regions() {
        // Return the union of regions in AWS_REGIONS and custom defined regions.
        return [...new Set([...aws_entities_1.AWS_REGIONS, ...Object.keys(this.database)])];
    }
    /**
     * Returns the list of names of registered facts.
     *
     * All facts will be present in at least one region.
     */
    static get names() {
        return [...new Set(Object.values(this.database).flatMap(regionFacts => Object.keys(regionFacts)))];
    }
    /**
     * Return all pairs of (region, factName) that are defined
     */
    static definedFacts() {
        return Object.entries(this.database)
            .flatMap(([regionName, regionFacts]) => Object.keys(regionFacts).map((factName) => [regionName, factName]));
    }
    /**
     * Retrieves a fact from this Fact database.
     *
     * @param region the name of the region (e.g: `us-east-1`)
     * @param name   the name of the fact being looked up (see the `FactName` class for details)
     *
     * @returns the fact value if it is known, and `undefined` otherwise.
     */
    static find(region, name) {
        const regionFacts = this.database[region];
        return regionFacts && regionFacts[name];
    }
    /**
     * Retrieve a fact from the Fact database. (retrieval will fail if the specified region or
     * fact name does not exist.)
     *
     * @param region the name of the region (e.g: `us-east-1`)
     * @param name the name of the fact being looked up (see the `FactName` class for details)
     */
    static requireFact(region, name) {
        const foundFact = this.find(region, name);
        if (!foundFact) {
            throw new RegionFactError(`No fact ${name} could be found for region: ${region} and name: ${name}.`);
        }
        return foundFact;
    }
    /**
     * Registers a new fact in this Fact database.
     *
     * @param fact           the new fact to be registered.
     * @param allowReplacing whether new facts can replace existing facts or not.
     */
    static register(fact, allowReplacing = false) {
        try {
            jsiiDeprecationWarnings._aws_cdk_region_info_IFact(fact);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.register);
            }
            throw error;
        }
        const regionFacts = this.database[fact.region] || (this.database[fact.region] = {});
        if (fact.name in regionFacts && regionFacts[fact.name] !== fact.value && !allowReplacing) {
            throw new RegionFactError(`Region ${fact.region} already has a fact ${fact.name}, with value ${regionFacts[fact.name]}`);
        }
        if (fact.value !== undefined) {
            regionFacts[fact.name] = fact.value;
        }
    }
    /**
     * Removes a fact from the database.
     * @param region the region for which the fact is to be removed.
     * @param name   the name of the fact to remove.
     * @param value  the value that should be removed (removal will fail if the value is specified, but does not match the
     *               current stored value).
     */
    static unregister(region, name, value) {
        const regionFacts = this.database[region] || {};
        if (name in regionFacts && value && regionFacts[name] !== value) {
            throw new RegionFactError(`Attempted to remove ${name} from ${region} with value ${value}, but the fact's value is ${regionFacts[name]}`);
        }
        delete regionFacts[name];
    }
    static database = {};
    constructor() {
        // this should never happen, so throw a regular error here
        /* eslint-disable-next-line @cdklabs/no-throw-default-error */
        throw new Error('Use the static methods of Fact instead!');
    }
}
exports.Fact = Fact;
/**
 * All standardized fact names.
 */
class FactName {
    static [JSII_RTTI_SYMBOL_1] = { fqn: "@aws-cdk/region-info.FactName", version: "2.235.0" };
    /**
     * The name of the partition for a region (e.g: 'aws', 'aws-cn', ...)
     */
    static PARTITION = 'partition';
    /**
     * The domain suffix for a region (e.g: 'amazonaws.com`)
     */
    static DOMAIN_SUFFIX = 'domainSuffix';
    /**
     * Whether the AWS::CDK::Metadata CloudFormation Resource is available in-region or not. The value is a boolean
     * modelled as `YES` or `NO`.
     */
    static CDK_METADATA_RESOURCE_AVAILABLE = 'cdk:metadata-resource:available';
    /**
     * Whether the given region is an opt-in region or not. The value is a boolean
     * modelled as `YES` or `NO`.
     */
    static IS_OPT_IN_REGION = 'aws:is-opt-in-region';
    /**
     * The endpoint used for hosting S3 static websites
     */
    static S3_STATIC_WEBSITE_ENDPOINT = 's3-static-website:endpoint';
    /**
     * The endpoint used for aliasing S3 static websites in Route 53
     */
    static S3_STATIC_WEBSITE_ZONE_53_HOSTED_ZONE_ID = 's3-static-website:route-53-hosted-zone-id';
    /**
     * The hosted zone ID used by Route 53 to alias a EBS environment endpoint in this region (e.g: Z2O1EMRO9K5GLX)
     */
    static EBS_ENV_ENDPOINT_HOSTED_ZONE_ID = 'ebs-environment:route-53-hosted-zone-id';
    /**
     * The prefix for VPC Endpoint Service names,
     * cn.com.amazonaws.vpce for China regions,
     * com.amazonaws.vpce otherwise.
     */
    static VPC_ENDPOINT_SERVICE_NAME_PREFIX = 'vpcEndpointServiceNamePrefix';
    /**
     * The account for ELBv2 in this region
     */
    static ELBV2_ACCOUNT = 'elbv2Account';
    /**
     * The ID of the AWS account that owns the public ECR repository that contains the
     * AWS Deep Learning Containers images in a given region.
     */
    static DLC_REPOSITORY_ACCOUNT = 'dlcRepositoryAccount';
    /**
     * The ID of the AWS account that owns the public ECR repository that contains the
     * AWS App Mesh Envoy Proxy images in a given region.
     */
    static APPMESH_ECR_ACCOUNT = 'appMeshRepositoryAccount';
    /**
     * The CIDR block used by Amazon Data Firehose servers.
     */
    static FIREHOSE_CIDR_BLOCK = 'firehoseCidrBlock';
    /**
     * The SAML Sign On URL for partition used by IAM SAML Principal
     */
    static SAML_SIGN_ON_URL = 'samlSignOnUrl';
    /**
     * The latest Lambda NodeJS runtime available in a given region.
     */
    static LATEST_NODE_RUNTIME = 'latestNodeRuntime';
    /**
     * The ARN of CloudWatch Lambda Insights for a version (e.g. 1.0.98.0)
     */
    static cloudwatchLambdaInsightsVersion(version, arch) {
        // if we are provided an architecture use that, otherwise
        // default to x86_64 for backwards compatibility
        const suffix = version.split('.').join('_') + `_${arch ?? 'x86_64'}`;
        return `cloudwatch-lambda-insights-version:${suffix}`;
    }
    /**
     * The ARN of AppConfig Lambda Layer for a given version (e.g. 2.0.181)
     * @param version The layer version.
     * @param arch The architecture (optional), defaults to x86_64.
     */
    static appConfigLambdaLayerVersion(version, arch) {
        const suffix = version.split('.').join('_') + `_${arch ?? 'x86_64'}`;
        return `appconfig-lambda-layer:${suffix}`;
    }
    /**
     * The name of the regional service principal for a given service.
     *
     * @param service the service name, either simple (e.g: `s3`, `codedeploy`) or qualified (e.g: `s3.amazonaws.com`).
     *                The `.amazonaws.com` and `.amazonaws.com.cn` domains are stripped from service names, so they are
     *                canonicalized in that respect.
     *
     * @deprecated - Use `iam.ServicePrincipal.servicePrincipalName()` instead.
     */
    static servicePrincipal(service) {
        try {
            jsiiDeprecationWarnings.print("@aws-cdk/region-info.FactName#servicePrincipal", "- Use `iam.ServicePrincipal.servicePrincipalName()` instead.");
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.servicePrincipal);
            }
            throw error;
        }
        return `${service.replace(/\.amazonaws\.com(\.cn)?$/, '')}.amazonaws.com`;
    }
    /**
     * The ARN of Amazon Distro for OpenTelemetry (ADOT) Lambda layer for a given lambda type, version and architecture.
     *
     * @param type the type of the ADOT lambda layer
     * @param version the layer version.
     * @param architecture the Lambda Function architecture (e.g. 'x86_64' or 'arm64')
     */
    static adotLambdaLayer(type, version, architecture) {
        const suffix = type + '_' + version.split('.').join('_') + '_' + architecture;
        return `adot-lambda-layer:${suffix}`;
    }
    /**
     * The ARN of Parameters and Secrets Lambda layer for a given lambda architecture.
     *
     * @param version the layer version
     * @param architecture the Lambda Function architecture (e.g. 'x86_64' or 'arm64')
     */
    static paramsAndSecretsLambdaLayer(version, architecture) {
        const suffix = version.split('.').join('_') + `_${architecture}`;
        return `params-and-secrets-layer:${suffix}`;
    }
}
exports.FactName = FactName;
